File: object/object/src/Lec02/Movie.java
package Lec02;

import Lec02.policy.discount.DiscountPolicy;
import Lec02.policy.refund.RefundPolicy;

import java.time.Duration;

/**
 * Movie 특: 예매하는 대상의 정보를 갖고 있다.
 *       책임 : 예매 금액 계산
 *
 */
public class Movie {

    private String title;
    private Duration runningTime;
    private Money fee;
    private DiscountPolicy discountPolicy;

    public Movie(String title, Duration runningTime, Money fee, DiscountPolicy discountPolicy) {
        this.title = title;
        this.runningTime = runningTime;
        this.fee = fee;
        this.discountPolicy = discountPolicy;
    }

    public Money getFee() {
        return fee;
    }

    public Money calculateMovieFee(Screening screening) {
        return fee.minus(discountPolicy.calculateDiscountAmount(screening));
    }

    public void changDiscountPolicy(DiscountPolicy discountPolicy) {
        this.discountPolicy = discountPolicy;
    }
}

File: object/object/src/java_to_kotlin/Movie.kt
package java_to_kotlin

import java_to_kotlin.policy.discount.DiscountPolicy
import java_to_kotlin.policy.refund.RefundPolicy
import java.time.Duration

/**
 * Movie : 예매하는 대상의 정보를 갖고 있다.
 * 책임 : 예매 대상 금액 계산
 *
 */
class Movie(private val title: String, private val runningTime: Duration, val fee: Money, private var discountPolicy: DiscountPolicy, private val refundPolicy: RefundPolicy) {
    fun calculateMovieFee(screening: Screening): Money? {
        return fee.minus(discountPolicy.calculateDiscountAmount(screening))
    }

    fun calculateRefundFee(refund: Refund): Money? {
        return refundPolicy.calculateRefundAmount(refund)
    }

    fun changDiscountPolicy(discountPolicy: DiscountPolicy) {
        this.discountPolicy = discountPolicy
    }
}
